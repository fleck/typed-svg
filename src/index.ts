import webpack from "webpack";
import { DOMParser } from "xmldom";
import fs from "fs-extra";
import eol from "eol";

export class TypedSVGWebpackPlugin {
  public apply(compiler: webpack.Compiler) {
    compiler.hooks.emit.tap(
      "CompileSvgSprite",
      (compilation: webpack.compilation.Compilation) => {
        const sprites = [
          ...compilation.fileDependencies,
        ].filter((fileName: string) => fileName.endsWith("sprite.svg"));

        sprites.forEach(filePath => {
          const domParser = new DOMParser();
          const doc = domParser.parseFromString(
            fs.readFileSync(filePath).toString(),
            "image/svg+xml"
          );

          const ids = Array.from(
            doc.getElementsByTagName("symbol")
          ).map(symbol => symbol.getAttribute("id"));

          fs.writeFileSync(
            `${filePath}.d.ts`,
            eol.lf(
              `declare const _default: string;
export default _default;

export const symbol: ${ids.map(id => `"${id}"`).join(" | ")};
`
            )
          );
        });
      }
    );
  }
}
