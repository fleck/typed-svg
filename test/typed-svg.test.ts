import webpack from "webpack";
import { TypedSVGWebpackPlugin } from "../src/index";
import path from "path";

it("Should generate types deceleration", async () => {
  let done: (value?: unknown) => void;

  const webpackFinished = new Promise(resolve => {
    done = resolve;
  });

  webpack(
    {
      plugins: [new TypedSVGWebpackPlugin()],
      entry: {
        main: path.join(__dirname, "addSvg.ts"),
      },
      resolve: {
        extensions: [".js", ".ts", ".svg"],
      },
      module: {
        rules: [
          {
            test: /\.(woff2|png|gif|svg|jpg)$/,
            use: {
              loader: "file-loader",
            },
          },
        ],
      },
    },
    (err, stats) => {
      if (err || stats.hasErrors()) {
        throw new Error(stats.toString());
      }
      done();
    }
  );

  await webpackFinished;
});
