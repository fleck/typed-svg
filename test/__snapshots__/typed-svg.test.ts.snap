// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should generate types deceleration 1`] = `
"// THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
declare const _default: string;
export default _default;

export const symbol: \\"first\\" | \\"second\\";
"
`;
